# 2. 🧹 Data cleaning
- In this notebook we are going to perform data cleaning operations to the `merged_df` dataset.

In [1]:
#lets import the libraries
import pandas as pd

In [2]:
#lets load the merged dataset
merged_df=pd.read_csv('C:/Users/Win/TMDB project/data/movies_full.csv')
merged_df.head()

,adult,backdrop_path,genre_ids,id,original_language,original_title,overview,popularity_x,poster_path,release_date_x,...,title_y,budget,revenue,runtime,status,release_date_y,popularity_y,vote_average_y,vote_count_y,genres
0,False,/7QirCB1o80NEFpQGlQRZerZbQEp.jpg,"[10749, 18]",1156594,es,Culpa nuestra,Jenna and Lion's wedding brings about the long...,855.1911,/yzqHt4m1SeY9FbPrfZ0C2Hi9x1s.jpg,2025-10-15,...,Our Fault,0,0,113,Released,2025-10-15,855.1911,7.803,338,"['Romance', 'Drama']"
1,False,/ygOR390GzOX5Quv0kAAcUNDG7fp.jpg,"[12, 28, 27]",1511789,en,Captain Hook - The Cursed Tides,In the aftermath of a devastating defeat by hi...,377.4300,/bcP7FtskwsNp1ikpMQJzDPjofP5.jpg,2025-07-11,...,Captain Hook - The Cursed Tides,0,0,90,Released,2025-07-11,377.4300,4.500,18,"['Adventure', 'Action', 'Horror']"
2,False,/2yo2k8CmgZxfBQYAZ8TnHIYzEme.jpg,"[28, 80, 53]",1072699,pl,Druga Furioza,"In the wake of murder, new Furioza leader Gold...",302.0473,/dbe1zKVKNBIczdmRcm7jxEphYLz.jpg,2025-10-14,...,Inside Furioza,0,0,166,Released,2025-10-14,302.0473,6.576,34,"['Action', 'Crime', 'Thriller']"
3,False,/iZLqwEwUViJdSkGVjePGhxYzbDb.jpg,"[878, 53]",755898,en,War of the Worlds,Will Radford is a top analyst for Homeland Sec...,312.4558,/yvirUYrva23IudARHn3mMGVxWqM.jpg,2025-07-29,...,War of the Worlds,0,0,91,Released,2025-07-29,312.4558,4.300,647,"['Science Fiction', 'Thriller']"
4,False,/qcDDn7WeKBenM4nLlOPXAeJ4hpg.jpg,"[53, 28, 12]",1280450,en,Stolen Girl,"In 1993, Maureen’s six-year-old daughter Amina...",294.0708,/fZlNXEHZsBp7unqw009MeBbMv87.jpg,2025-09-04,...,Stolen Girl,26000000,92691,110,Released,2025-09-04,294.0708,6.686,35,"['Thriller', 'Action', 'Adventure']"


In [3]:
merged_df.columns

Index(['adult', 'backdrop_path', 'genre_ids', 'id', 'original_language',
       'original_title', 'overview', 'popularity_x', 'poster_path',
       'release_date_x', 'title_x', 'video', 'vote_average_x', 'vote_count_x',
       'title_y', 'budget', 'revenue', 'runtime', 'status', 'release_date_y',
       'popularity_y', 'vote_average_y', 'vote_count_y', 'genres'],
      dtype='object')

In [4]:
#first we  are going to drop all the x-labeled columns since they appeared because both dataframes had overlapping column names
# they basically have the same info
merged_df=merged_df.drop(columns=['popularity_x', 'release_date_x', 'title_x', 'vote_average_x', 'vote_count_x'])
merged_df.columns

Index(['adult', 'backdrop_path', 'genre_ids', 'id', 'original_language',
       'original_title', 'overview', 'poster_path', 'video', 'title_y',
       'budget', 'revenue', 'runtime', 'status', 'release_date_y',
       'popularity_y', 'vote_average_y', 'vote_count_y', 'genres'],
      dtype='object')

In [5]:
#lets noew rename the y_labeled columns to cleaner names
merged_df=merged_df.rename(columns={
    'title_y':'title',
    'popularity_y':'popularity',
    'release_date_y':'release_date',
    'vote_average_y':'vote_average',
    'vote_count_y':'vote_count'
    
})
merged_df.columns

Index(['adult', 'backdrop_path', 'genre_ids', 'id', 'original_language',
       'original_title', 'overview', 'poster_path', 'video', 'title', 'budget',
       'revenue', 'runtime', 'status', 'release_date', 'popularity',
       'vote_average', 'vote_count', 'genres'],
      dtype='object')

In [6]:
#lets see the shape of the dataset
merged_df.shape
print("shape of the dataset",merged_df.shape)


shape of the dataset (4218, 19)


In [7]:
#lets get some information on the dataset
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4218 entries, 0 to 4217
Data columns (total 19 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   adult              4218 non-null   bool   
 1   backdrop_path      4118 non-null   object 
 2   genre_ids          4218 non-null   object 
 3   id                 4218 non-null   int64  
 4   original_language  4218 non-null   object 
 5   original_title     4218 non-null   object 
 6   overview           4193 non-null   object 
 7   poster_path        4210 non-null   object 
 8   video              4218 non-null   bool   
 9   title              4218 non-null   object 
 10  budget             4218 non-null   int64  
 11  revenue            4218 non-null   int64  
 12  runtime            4218 non-null   int64  
 13  status             4218 non-null   object 
 14  release_date       4213 non-null   object 
 15  popularity         4218 non-null   float64
 16  vote_average       4218 

In [8]:
#lets get a statistical description of our dataset
merged_df.describe()

,id,budget,revenue,runtime,popularity,vote_average,vote_count
count,4.218000e+03,4.218000e+03,4.218000e+03,4218.00000,4218.000000,4218.000000,4218.000000
mean,4.317092e+05,3.500275e+07,1.211736e+08,105.20294,10.684524,6.472386,3245.167378
std,4.724833e+05,5.584956e+07,2.317135e+08,29.33841,21.784320,1.459040,4720.394758
min,1.100000e+01,0.000000e+00,0.000000e+00,0.00000,3.787100,0.000000,0.000000
25%,1.133450e+04,0.000000e+00,0.000000e+00,92.00000,5.912325,6.066500,105.000000
50%,2.506400e+05,9.000000e+06,2.048841e+07,104.00000,7.056250,6.688000,1339.000000
75%,8.040202e+05,4.800000e+07,1.453701e+08,119.00000,9.585000,7.300000,4447.250000
max,1.565344e+06,5.839000e+08,2.923706e+09,400.00000,855.191100,10.000000,38099.000000


In [9]:
#lets check for duplicates
merged_df.duplicated().sum()

np.int64(326)

- **we have 326 duplicated rows in our dataset, we opt to drop them**

In [10]:
#lets drop the duplicates
merged_df.drop_duplicates(inplace=True)

In [11]:
#lets confirm there are no more duplicates
merged_df.duplicated().sum()

np.int64(0)

In [12]:
#lets check for null values
merged_df.isnull().sum()

adult                 0
backdrop_path        94
genre_ids             0
id                    0
original_language     0
original_title        0
overview             25
poster_path           8
video                 0
title                 0
budget                0
revenue               0
runtime               0
status                0
release_date          5
popularity            0
vote_average          0
vote_count            0
genres                0
dtype: int64

In [13]:
merged_df.head(2)

,adult,backdrop_path,genre_ids,id,original_language,original_title,overview,poster_path,video,title,budget,revenue,runtime,status,release_date,popularity,vote_average,vote_count,genres
0,False,/7QirCB1o80NEFpQGlQRZerZbQEp.jpg,"[10749, 18]",1156594,es,Culpa nuestra,Jenna and Lion's wedding brings about the long...,/yzqHt4m1SeY9FbPrfZ0C2Hi9x1s.jpg,False,Our Fault,0,0,113,Released,2025-10-15,855.1911,7.803,338,"['Romance', 'Drama']"
1,False,/ygOR390GzOX5Quv0kAAcUNDG7fp.jpg,"[12, 28, 27]",1511789,en,Captain Hook - The Cursed Tides,In the aftermath of a devastating defeat by hi...,/bcP7FtskwsNp1ikpMQJzDPjofP5.jpg,False,Captain Hook - The Cursed Tides,0,0,90,Released,2025-07-11,377.4300,4.500,18,"['Adventure', 'Action', 'Horror']"


In [14]:
#lets drop some columns since they are of no significance in our project
merged_df.drop(columns=['backdrop_path','poster_path','adult', 'video', 'status',
                        'original_title', 'id','genre_ids'],inplace=True)

#WHY WE DROPPED:
#adult: almost all are False, no predictive value
#video: always False for most movies
#status: usually “Released” for all
#overview:long text, unless you plan NLP analysis
#original_title: duplicate of title
#id: identifier only, 
#genre_ids: just a list of numeric lists reperesenting the genres in the genre column

In [15]:
merged_df.columns

Index(['original_language', 'overview', 'title', 'budget', 'revenue',
       'runtime', 'release_date', 'popularity', 'vote_average', 'vote_count',
       'genres'],
      dtype='object')

In [16]:
#lets drop the five rows with missing release_date
merged_df.dropna(subset=['release_date'], inplace=True)

In [17]:
#lets also drop the rows with missing overview columns
merged_df.dropna(subset=['overview'], inplace=True)

In [18]:
merged_df.isnull().sum()

original_language    0
overview             0
title                0
budget               0
revenue              0
runtime              0
release_date         0
popularity           0
vote_average         0
vote_count           0
genres               0
dtype: int64

- **🙌 now we have no missing values**
- **LETS DEFINE THE COLUMNS**

- original_language → language the movie was produced in (e.g., "en", "fr")

- overview → short plot summary or description of the movie

- title → movie’s official title

- budget → production cost (in USD)

- revenue → total earnings (in USD) — your target variable

- runtime → movie duration in minutes

- release_date → when the movie was released

- popularity → TMDB’s popularity metric based on views and ratings

- vote_average → average user rating (0–10)

- vote_count → total number of user ratings

- genres → list of movie genres (e.g., Action, Drama, Comedy)

In [19]:
#since we can observe that, some row have zero entries in the revenue an budget columns , lets see how many
print("zero entries in the revenue column:",(merged_df['revenue'] == 0).sum())
print("zero entries in the budget column:",(merged_df['budget'] == 0).sum())


zero entries in the revenue column: 1328
zero entries in the budget column: 1432


In [20]:
# we opt to drop these rows for practicality
merged_df = merged_df[(merged_df['budget'] != 0) & (merged_df['revenue'] != 0)]


In [21]:
print("The shape of our dataset after cleaning",merged_df.shape)

The shape of our dataset after cleaning (2268, 11)


In [22]:
merged_df.head()

,original_language,overview,title,budget,revenue,runtime,release_date,popularity,vote_average,vote_count,genres
4,en,"In 1993, Maureen’s six-year-old daughter Amina...",Stolen Girl,26000000,92691,110,2025-09-04,294.0708,6.686,35,"['Thriller', 'Action', 'Adventure']"
5,ja,The Demon Slayer Corps are drawn into the Infi...,Demon Slayer: Kimetsu no Yaiba Infinity Castle,20000000,643612593,156,2025-07-18,261.8345,7.793,474,"['Animation', 'Action', 'Fantasy', 'Thriller']"
6,kn,Martin revolves around Lt. Brigadier Arjun Sax...,Martin,17647000,3247000,147,2024-10-11,271.6580,2.100,8,"['Action', 'Drama', 'Thriller']"
8,en,Paranormal investigators Ed and Lorraine Warre...,The Conjuring: Last Rites,55000000,482377782,135,2025-09-03,238.2160,7.000,1080,['Horror']
11,en,When The Strangers learn that one of their vic...,The Strangers: Chapter 2,8500000,11577352,98,2025-09-25,150.7366,5.850,70,"['Horror', 'Thriller']"


- **Even though the dataset shrinked,We now have a clean dataset.**

In [23]:
#lets now save the cleaned dataset
merged_df.to_csv('C:/Users/Win/TMDB project/data/cleaned_movies.csv', index=False)


## NEXT WE MOVE TO EDA(Exploratory Data Analysis)